### PR TITLE
CI: On vX.Y-rhel branches, ensure that some downstream Jira issue is linked

### PR DESCRIPTION
**Pull request by alexlarsson**
In the RHEL specific branches we want to ensure that all MRs link to at least one downstream Jira ticket. To do this we add a new test in validate-source similar to the existing pr-should-include-tests. This test only runs on actual pull requests.
    
The syntax for linking to a Jira is "Fixes " followed by one jira links, like so:  
```
Fixes https://issues.redhat.com/browse/RHEL-50506
```
    
Note: This is the same syntax as for a regular github issue reference.

### CODE REVIEW

**Comment by alexlarsson** [comment]
@baude @mheon @TomSweeneyRedHat @edsantiago follow up from the email discussion.

The exact syntax, as well as the label name for the override is just an initial proposal for me, and is up for discussion.

**Comment by alexlarsson** [comment]
Btw, the DEST_BRANCH in .cirrus.yml on the "v4.9-rhel" branch is wrong (it is "v4.9"), and we have to update it if we want this to work there.

**Comment by mheon** [comment]
I'm thinking we're going to mandate all PRs on upstream that fix an issue reference that issue as "Fixes <issue>" where issue could be a Github issue ID, a URL, or a Jira reference. So I think matching "Fixes Jira <key>-<number>\n" would be preferable. But the overall approach LGTM.
This will need to be backported to each -rhel branch, we take CI from per-branch config not main.

**Comment by mheon** [comment]
@cevich Is there any way we can enforce Cirrus DEST_BRANCH being set correctly, because I think we're especially bad about remembering to do this for -rhel branches?

**Comment by alexlarsson** [comment]
Maybe we can have a github action (not in cirrus) that validates the cirrus DEST_BRANCH value? I think you can get that from $GITHUB_BASE_REF

**Comment by cevich** [comment]
> Maybe we can have a github action (not in cirrus) that validates the cirrus DEST_BRANCH value? I think you can get that from $GITHUB_BASE_REF

That's not a bad suggestion actually since github actions on this repo can send e-mail to our monitoring list.

OTOH, the base-branch name _is_ exposed in a Cirrus-CI env. var. (can't recall the exact name) to PR-level tasks but nowhere else (i.e. not in tag or branch-level jobs).  So maybe a simple check could be done _if_ `$CIRRUS_BASE_BRANCH != main && $CIRRUS_BASE_BRANCH != $DEST_BRANCH` or something to that effect.

**Review by Luap99 — CHANGES_REQUESTED** [review]
Not against this per se, just marking that I don't think we should merge this until we have a general accept format defined and documented before we "enforce" anything

**Comment by mheon** [comment]
@Luap99 I'm going to proposal the generic `Fixes [issue]` formatting on Monday at standup.

File: contrib/cirrus/pr-should-link-jira
Comment (TomSweeneyRedHat) [inline contrib/cirrus/pr-should-link-jira:13]: Don't we need main also?  That eventually turns into a Jira branch.

File: contrib/cirrus/pr-should-link-jira
Comment (TomSweeneyRedHat) [inline contrib/cirrus/pr-should-link-jira:14]: Only Podman has the `-rhel` branch thought.  If you want to do other branches, this will have to be adjusted somehow.

**Review by TomSweeneyRedHat — COMMENTED** [review]
This needs to be documented somwehere.  Perhaps https://github.com/containers/podman/blob/main/CONTRIBUTING.md

File: contrib/cirrus/pr-should-link-jira
Comment (TomSweeneyRedHat) [inline contrib/cirrus/pr-should-link-jira:73]: This would be great in CONTRIBUTING.md too

File: contrib/cirrus/pr-should-link-jira
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira:13]: Not sure what you mean by that. There are tons of changes in main that don't have a redhat jira issue to link to?

File: contrib/cirrus/pr-should-link-jira
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira:14]: Not sure what kind of branches you mean, care to give an example?

File: contrib/cirrus/pr-should-link-jira
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira:73]: Do we want to add redhat specific stuff there? Maybe in a separate section.

Comment (Luap99) [reply contrib/cirrus/pr-should-link-jira:73]: If this is a RHEL only process it should not be in CONTRIBUTING.md as a normal contributor should not ever need to deal with Jira. I do however agree that we need a general documentation how we link issues/fixes in commits.

**Comment by alexlarsson** [comment]
> @Luap99 I'm going to proposal the generic `Fixes [issue]` formatting on Monday at standup.

Any chance I can join that?

**Comment by mheon** [comment]
@alexlarsson Sure, I can invite you.

**Comment by alexlarsson** [comment]
@mheon Updated the PR

**Comment by mheon** [comment]
Someone who is better at regexes than I should review, but LGTM

File: contrib/cirrus/pr-should-link-jira.t
Comment (edsantiago) [inline contrib/cirrus/pr-should-link-jira.t:26]: This needs to be closer to line 64, where the test table is defined

File: contrib/cirrus/pr-should-link-jira.t
Comment (edsantiago) [inline contrib/cirrus/pr-should-link-jira.t:73]: Could you table-align the branch names please? This is very difficult for a human to read

File: contrib/cirrus/pr-should-link-jira.t
Comment (edsantiago) [inline contrib/cirrus/pr-should-link-jira.t:65]: Why not use a real PR? Like #23514 and #23567? Even better, those plus #8890, so you can add this test:
```
0  8890 msg_no_jira  v1.6.4-rhel     no link, but PR has magic label, should pass
```
(I have taken the liberty of creating a `No Jira Link` label and applied it to this very ancient PR).

**Review by edsantiago — COMMENTED** [review]
Good start, but a few suggestions inline

File: contrib/cirrus/pr-should-link-jira.t
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira.t]: done

File: contrib/cirrus/pr-should-link-jira.t
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira.t]: done

File: contrib/cirrus/pr-should-link-jira.t
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira.t]: I updated the table to use #8890 to test getting the tag. However, I'm not sure what you mean by using the others for the checks, as we're currently embedding the tested pr descriptions in the test itself, we're not getting it from github. The github PR is only used for checking the override tag atm.

Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira.t]: I made them 0000 for the non-queried ones, and #23514 for the ones we query that have no label. Does that look better?

Comment (edsantiago) [reply contrib/cirrus/pr-should-link-jira.t]: Better, thank you! LGTM.

File: contrib/cirrus/pr-should-link-jira
Comment (cevich) [inline contrib/cirrus/pr-should-link-jira:16]: Off-hand, I'm not sure bash understands `\w', and I think you probably want `\\w` no?

File: contrib/cirrus/pr-should-link-jira
Comment (cevich) [inline contrib/cirrus/pr-should-link-jira:24]: This will only match the HEAD commit.  Maybe as a followup, consider checking all commits to a PR?

File: contrib/cirrus/pr-should-link-jira
Comment (cevich) [inline contrib/cirrus/pr-should-link-jira:27]: This variable will be empty for executions on branches and tags, causing this script to fail and needlessly fire off monitoring alert e-mails.  I think maybe an easy fix is to simply set it to `exit 0` instead?

File: contrib/cirrus/pr-should-link-jira
Comment (cevich) [inline contrib/cirrus/pr-should-link-jira:53]: I'm not certain about the github API behavior, what happens if there are no labels at all?  Consider if something like this might be more robust:

```suggestion
labels=$(jq -r '.data.repository.pullRequest.labels.nodes[]?.name' <<<"$result")
```

File: contrib/cirrus/pr-should-link-jira.t
Comment (cevich) [inline contrib/cirrus/pr-should-link-jira.t:1]: To ensure the script doesn't break in the future, it would be good to execute the tests as part of CI.  Right after the `lib.sh.t` test would be a great place:

https://github.com/containers/podman/blob/main/contrib/cirrus/prebuild.sh#L62-L64

File: contrib/cirrus/pr-should-link-jira
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira:24]: No, for a PR, CIRRUS_CHANGE_MESSAGE contains the PR description, not a commit message, and this test only runs on PRs. And, this is what we want, as then you can easily edit the description instead of having to modify a commit.

File: contrib/cirrus/pr-should-link-jira
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira:27]: This test is called from _run_validate-source, which is only run during PRs,, as per:
```
validate-source_task:
    name: "Validate source code changes"
    alias: validate-source
    # This task is primarily intended to catch human-errors early on, in a
    # PR context.  Skip running it everywhere else.
    only_if: &is_pr "$CIRRUS_PR != ''"
```
So, this is not a problem. 
In fact, this code is an identical copy from the similar pr-should-include-tests that already works for the same reason.

File: contrib/cirrus/pr-should-link-jira
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira:53]: This code was also copied from pr-should-include-tests, so I think it works. But I can change this if you want.

File: contrib/cirrus/pr-should-link-jira
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira:16]: I did some experimenting with $BASH_REMATCH to see what was actually matched, and you are right, its not supported. So I rewrote it to use `[:alnum:]` style matchers instead, which works nicely.

File: contrib/cirrus/pr-should-link-jira
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira]: I changed it

File: contrib/cirrus/pr-should-link-jira.t
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira.t:1]: Ah, no it fails because GITHUB_TOKEN is not set

Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira.t:1]: Fixed it, should work now.

File: contrib/cirrus/pr-should-link-jira.t
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira.t]: Pushed, lets see if that works.

File: contrib/cirrus/pr-should-link-jira.t
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira.t:18]: I added it there so i could test it with:
```
CIRRUS_REPO_CLONE_TOKEN=... ./pr-should-link-jira.t
```
Otherwise it would not get propagated to the test.
That said, one might very well not want to support that...

File: contrib/cirrus/pr-should-link-jira
Comment (cevich) [reply contrib/cirrus/pr-should-link-jira:24]: Gotcha, sorry I assumed it was looking at commit messages.

File: contrib/cirrus/pr-should-link-jira
Comment (cevich) [reply contrib/cirrus/pr-should-link-jira:27]: You are correct, though scripts and tasks do get shuffled around now and again.  Would you consider:

```bash
if [[ -z "$CIRRUS_PR" ]]; then
    ...
elif [[ ! "$CIRRUS_BRANCH" =~ pull ]] || [[ -n "$CIRRUS_TAG" ]]; then
    echo "Warning: $ME only intended for use on PRs"
    exit 0
fi
```

File: contrib/cirrus/pr-should-link-jira.t
Comment (cevich) [reply contrib/cirrus/pr-should-link-jira.t:18]: If you do want to keep this here, I would be helpful to add a comment suggesting it's needed for pr-should-link-jira.t to function.

**Comment by alexlarsson** [comment]
The mac build always seems to time out on the linter. I assume that is not my fault?

**Comment by cevich** [comment]
> The mac build always seems to time out on the linter. I assume that is not my fault?

I think there was a fix for that recently and rebasing on main should fix it.  I might be wrong though.

**Comment by alexlarsson** [comment]
> > The mac build always seems to time out on the linter. I assume that is not my fault?
> 
> I think there was a fix for that recently and rebasing on main should fix it. I might be wrong though.

I tried rebasing, lets see.

File: contrib/cirrus/pr-should-link-jira.t
Comment (alexlarsson) [reply contrib/cirrus/pr-should-link-jira.t]: I moved it.

**Comment by edsantiago** [comment]
Still LGTM

**Comment by alexlarsson** [comment]
Ok, all requested changes are done, and now we passed CI. Can I get this merged?

**Comment by alexlarsson** [comment]
(Note: Once this is merged we need to backport to active rhel branches.

**Comment by mheon** [comment]
That might have confused the bot
/lgtm

**Pull request merged**
